python/tskit/tables.py

```diff
@@ -2470,3 +2470,172 @@ def drop_index(self):
         currently indexed this method has no effect.
         """
         self.ll_tables.drop_index()
+
+    def add_time(self, dt):
+        """
+        This function modifies a table collection IN PLACE by adding ``dt`` to
+        all times in the node table.
+        """
+        nodes_dict = self.nodes.asdict()
+        nodes_dict['time'] = nodes_dict['time'] + dt
+        self.nodes.set_columns(**nodes_dict)
+        migrations_dict = self.migrations.asdict()
+        migrations_dict['time'] = migrations_dict['time'] + dt
+        self.migrations.set_columns(**migrations_dict)
+
+    def graft(self, other, node_map, check_overlap=False):
+        """
+        Returns tables by grafting the table collection ``other`` onto this
+        table collection along the nodes specified in the dictionary ``node_map``,
+        which should be a dictionary whose keys are node IDs in ``other`` and values
+        are the equivalent node IDs in this set of tables.
+        Any populations of nodes that are newly added to this table collection
+        are considered new in the grafted tables, even if their IDs are the same.
+        Map is returned (TODO: say exactly what).
+        If ``check_overlap`` is True, then check whether the two tables, simplified
+        to the shared nodes, are equal. This check is optional because it is relatively
+        expensive. TODO: should it be optional?
+        """
+        # checking shift in time ago between self and other
+        dt = [self.nodes.time[j] - other.nodes[i].time for i, j in node_map.items()]
```

Comment:
There's room for numerical imprecision errors here I'd imagine. Probably need to use numpy almost equal.

Reply:
right, I didn't think carefully about this. also not sure how to do the add_time operation when we can have numerical imprecision?

Reply:
Nothing special in `add_time`, I don't think.

Reply:
it is harder to define the amount to add, no?

Reply:
no? we're still allowed to subtract floats from each other?

Reply:
well, there won't be one unique time difference to add. is it OK to just choose the first time difference provided the times in self are close enough to the times in other?

Reply:
Yes, that's right - they'll all be equal up to machine precision, which is what we care about.

Reply:
ok, I removed it from the graft method